utils/utils.c

```diff
@@ -154,11 +155,19 @@ int remove_directory(char *dirname)
 			continue;
 
 		snprintf(buf, sizeof(buf), "%s/%s", dirname, ent->d_name);
-		if (unlink(buf) < 0) {
+		if (stat(buf, &statbuf)) {
+failed:
 			saved_errno = errno;
 			ret = -1;
 			break;
 		}
+
+		if (S_ISDIR(statbuf.st_mode)) {
+			if (remove_directory(buf) < 0)
+				goto failed;
+		}
+		else if (unlink(buf) < 0)
+			goto failed;
```

Comment:
I'd like to avoid backward jump (goto) if possible.  Could you please just save the result value and check it at the end of the loop body?

Reply:
Yep, I got it !

Reply:
I changed this PR again !
I'd appreciate it, if you check this 😄 

Additionally if you don't want nested if statement,
change it like below ?
```patch
diff --git a/utils/utils.c b/utils/utils.c
index 67b1903..254dcae 100644
--- a/utils/utils.c
+++ b/utils/utils.c
@@ -138,6 +138,7 @@ int remove_directory(char *dirname)
 {
         DIR *dp;
         struct dirent *ent;
+        struct stat statbuf;
         char buf[PATH_MAX];
         int saved_errno = 0;
         int ret = 0;
@@ -154,9 +155,21 @@ int remove_directory(char *dirname)
                         continue;
 
                 snprintf(buf, sizeof(buf), "%s/%s", dirname, ent->d_name);
-                if (unlink(buf) < 0) {
-                        saved_errno = errno;
+                if (stat(buf, &statbuf)) {
                         ret = -1;
+                        goto failed;
+                }
+
+                if (S_ISDIR(statbuf.st_mode)) {
+                        if (remove_directory(buf) < 0)
+                                ret = -1;
+                }
+                else if (unlink(buf) < 0)
+                        ret = -1;
+
+                if (ret < 0) {
+failed:
+                        saved_errno = errno;
                         break;
                 }
         }

```

Reply:
What about this?

```c
   ret = stat();
   if (ret < 0)
      goto failed;

   if (S_ISDIR(...))
      ret = remove_directory();
   else
      ret = unlink();

   if (ret < 0) {
failed:
       saved_errno = errno;
       break;
   }
```

Reply:
That is better !
I changed this PR based on your code,
so I added your signed-off-by to commit log message. 😄